AY2223S1-CS2103-F14-1/tp
# `nusmods` API interface

Closes #36

## Description

Data models and http client is generated by [openapi-generator](https://openapi-generator.tech/), `swagger-codegen` was abandoned because of its lack of support for `openapi 3.X.X` (cannot handle `oneof` relationships well)

Command used to generate client:

```bash
openapi-generator generate -i https://api.nusmods.com/v2/swagger.yaml -g java -o codegen --additional-properties=bigDecimalAsString=true --additional-properties=library=native
```

`openapi` schema from [nusmods api](https://api.nusmods.com/v2/) `src/main/java/org/openapitools/client/api/openapi.yaml` (kept in repo for reference)

## Problems faced

### `List<BigDecimal>`

Their generated client was not able to handle `List<BigDecimal>` well which is the data type used to encode `WorkLoad` and `LessonWeeks`, so I had to change the deserializer a little.

https://github.com/AY2223S1-CS2103-F14-1/tp/blob/3e727c2ccc94f3f4dee2b63f7af76d6f2e2e1170/src/main/java/org/openapitools/client/model/Workload.java#L85-L96

I simply used the `List` construct to do deserialization instead of `List<BigDecimal>` which will result in a compile time error since you cannot do `List<BigDecimal>.class`

> Strategy is the same for `LessonWeeks`

## How to test

I have added a simple `junit` test as a sanity check that the `openapi` client works. Do try to run it locally and see if the tests pass. Feel to suggest more tests to add as well :)

It is under `src/test/java/org/openapitools/client/api/ModulesApiTest.java`
